### ISSUE TITLE
[v3][Bug] useFormValue is out of sync

### ISSUE THREAD

**Issue by Petlas88** [sanity-io/sanity#3959]
If you find a security vulnerability, do NOT open an issue. Email [security@sanity.io](mailto:security@sanity.io) instead.

**Describe the bug**

The `useFormValue` hook used to access document values is out of sync (always one step behind). In my specific case it is a custom slug filed that should update paths based on category and parent document that is set from other field. In this case, if I click a category, nothing updates in the slug field the first time. But the second time I click you can see it briefly set the first category I clicked, before setting the last category. So in short it skips one click.

When refreshing the browser the values are correct, but if someone publishes the document without refreshing, the document will be published with the wrong values for the slug.

In case of regular string inputs, the `useFormValue` hook is always one character behind, which can cause issues if, say one field should have some special functionality if one document value is empty.

**To Reproduce**

- Create an input field like this (taken from [here](https://www.sanity.io/docs/migration-cheat-sheet)):

```
import {useFormValue} from 'sanity'

function MyInput() {
  const title = useFormValue([`title`])
  return (
    <div>
      Document title: {title}
    </div>
  )
}

export default MyInput
```

- Create a schema like this: 
```
export default defineType({
  name: "test",
  title: "Test",
  type: "document",
  fields: [
    defineField({
      name: "title",
      title: "Title",
      type: "string",
    }),
    defineField({
      name: "myInput",
      title: "My Input",
      type: "string",
      components: { input: MyInput },
    }),
  ],
});
```

- Add a new document, start typing in the title field. The custom component should be one character behind.

**Expected behavior**

The `useFormValue` hook should update in sync with document changes, like props when using `withDocument` did in v2.

**Video**

https://user-images.githubusercontent.com/42611458/207545769-9a5e28a4-5883-4b8d-be1a-86e35f69af32.mov

**Which versions of Sanity are you using?**

```
@sanity/cli (global)           3.0.6 (up to date)
@sanity/base                  2.35.2 (up to date)
@sanity/cli                    3.0.6 (up to date)
@sanity/dashboard              3.0.0 (up to date)
@sanity/default-layout        2.35.2 (up to date)
@sanity/default-login         2.35.2 (up to date)
@sanity/desk-tool             2.35.2 (up to date)
@sanity/eslint-config-studio   2.0.1 (up to date)
@sanity/production-preview    2.33.2 (up to date)
@sanity/scheduled-publishing   1.0.2 (latest: 1.1.0)
@sanity/vision                 3.0.6 (up to date)
sanity                         3.0.6 (up to date)
```

**What operating system are you using?**
```
MacOS Monterey 12.5.1 
```

**Which versions of Node.js / npm are you running?**

Node `v16.15.1`
Npm `8.11.0`

**Comment by Aratramba** [comment]
is this maybe the same as https://github.com/sanity-io/sanity/issues/3606 ?

**Comment by Petlas88** [comment]
> is this maybe the same as #3606 ?

Looks like it's the same issue, yes.

**Comment by aascione-mw** [comment]
I am migrating an existing solution from v2 and I am experiencing the same issue.

**Comment by dutsik** [comment]
The same thing here. we are migrating existing solution from v2 to v3. and experience issues with custom Fields that depends on current document state. For instance nested dropdowns, when you select a category and then the subcategory of that category.

I believe that it is a quite common use case. is there any estimates on when it will be fixed?

**Comment by dutsik** [comment]
Guys is there any workaround?

**Comment by Petlas88** [comment]
> Guys is there any workaround?

I tried using [useListeningQuery](https://github.com/SimeonGriggs/sanity-plugin-utils/blob/main/src/hooks/useListeningQuery.tsx), which kind of works, but it's also a bit out of sync it seemed. The main problem though, is that it will eat your monthly API requests if you have a cap. This made us not be able to use it. But it might work for you. 

We have to redo a lot of what we've done in v3  in v2 as a temporary solution now. Which is a little uncool as we have everything ready to go in v3. This issue is the one thing keeping us from deploying v3, and causes double work as we need to maintain the v2 one with new features/schemas even though everything is ready to go in v3.

**Comment by dutsik** [comment]
Not sure if this is the correct way or not, but.
Here is what I have found:
use
```
useFormBuilder().value[parentFacetPropertyName]
```
instead of:
```
useFormValue([parentFacetPropertyName])
```
and it did the trick. :)

**Comment by Petlas88** [comment]
@dutsik this worked for me as well! Probably not the way it should be, but it works until useFormValue is fixed!. Thank you so very much for this! 😄

**Comment by bjoerge** [comment]
Fixed by #4069 and released with [v3.2.0](https://github.com/sanity-io/sanity/releases/tag/v3.2.0)
